### PR TITLE
Add dual layouts for Draco_Mesh vertices to local and ghost cells.

### PR DESCRIPTION
**Pull request by RyanWollaeger**
### Background

* Certain spatial stencils require corner neighbor data, or data from cells only connected via nodes (or vertices) not constituting a full face.
* In an unstructured mesh, any number of off-MPI rank or ghost cells may intersect a local cell by a subset of nodes.
  * In 2D, this intersection would be one node.
  * In 3D, this intersection can be any number less than the number on a face.
* The ghost cell data in the constructor of `Draco_Mesh` only takes ghost data for off-rank cell faces, but the additional connectivity between cells via nodes can be generated with some additional MPI calls.

### Purpose of Pull Request

* Add dual-layouts to `Draco_Mesh` for local cells connected via (at least) nodes and for nodes connecting to off-rank cells.
* These layouts are maps from local node indices to a vector of connecting local cells (`Dual_Layout`) or a vector of pairs of off-rank cells and their corresponding ranks (`Dual_Ghost_Layout`).
* *This PR is marked as WIP pending more testing and review, as well as consideration if this dual layout capability should be optional.*

### Description of changes

+ Add Dual_Layout and Dual_Ghost_Layout types to Draco_Mesh.hh.
+ Add function to compute node-cell linkage and ghost linkage.
+ Use this new function in Draco_Mesh ctor after main layout gen.
+ Use c4/MPI functions for all-to-all gather of ghost node data.
+ Add MPI_Allgatherv versions of some c4 functions.
+ Test node-(ghost-)cell layouts tstDraco_Mesh(_DD) unit tests.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: src/c4/C4_MPI.t.hh [hunk 1]
```diff
@@ -230,6 +230,17 @@ int gatherv(T *send_buffer, int send_size, T *receive_buffer, int *receive_sizes
   return Result;
 }
 
+//------------------------------------------------------------------------------------------------//
+template <typename T>
```
Comment (alexrlongne) [inline src/c4/C4_MPI.t.hh:234]: Could you add a doxygen comment block here? Thanks!

File: src/c4/gatherv.t.hh [hunk 2]
```diff
@@ -166,6 +166,114 @@ void determinate_gatherv(std::vector<T> &outgoing_data,
   return;
 }
 
+//------------------------------------------------------------------------------------------------//
+template <typename T>
+void indeterminate_allgatherv(std::vector<T> &outgoing_data,
```
Comment (alexrlongne) [inline src/c4/gatherv.t.hh:171]: Could you also a comment block here? Sorry, I'm not very familiar with these types of communication patterns.

File: src/c4/C4_MPI.t.hh (hunk 1 above)
Comment (KineticTheory) [reply src/c4/C4_MPI.t.hh:234]: You will also need to add a stub with this signature to `C4_Serial.t.hh`.

File: src/mesh/Draco_Mesh.cc [hunk 3]
```diff
@@ -433,6 +440,220 @@ Draco_Mesh::compute_node_vec_indx_map(const std::vector<unsigned> &indx_type,
   return nodes_to_indx_map;
 }
 
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief Build a cell-to-cell linkage across corners
+ *
+ */
```
Comment (KineticTheory) [inline src/mesh/Draco_Mesh.cc:447]: Please add `\param` for this function.

**Review by KineticTheory — APPROVED** [review]
There are some regression failures that need to be fixed.  Otherwise, LGTM.

File: src/c4/C4_MPI.t.hh (hunk 1 above)
Comment (RyanWollaeger) [reply src/c4/C4_MPI.t.hh:242]: Sorry for the delay - will get these modifications in shortly.

Comment (RyanWollaeger) [reply src/c4/C4_MPI.t.hh:242]: @alexrlongne Added a doxygen comment block to the corresponding place in `C4_Functions.hh` (it seems this is where the other comment blocks are).
@KineticTheory Added a serial stub version of this function in `C4_Serial.t.hh`.

File: src/c4/gatherv.t.hh (hunk 2 above)
Comment (RyanWollaeger) [reply src/c4/gatherv.t.hh:171]: Added full comment blocks to the header `gatherv.hh` and updated the existing ones to (hopefully) be more descriptive.

File: src/mesh/Draco_Mesh.cc (hunk 3 above)
Comment (RyanWollaeger) [reply src/mesh/Draco_Mesh.cc:445]: Thank you for catching this, these comment fields have been added.

**Comment by RyanWollaeger** [comment]
Also added in a recent commit is a 4-rank test, which has a less trivial test of corner-data connectivity across ranks only sharing one mesh vertex.

**Comment by RyanWollaeger** [comment]
Removed WIP status.

**Comment by KineticTheory** [comment]
@RyanWollaeger Thanks for the updates.  The Travis CI has been painfully slow.  We might need to wait until Wed. to merge this.
